File: compiler/bench/life.c
// expected: 5
// Game of Life - run simulation steps
#include <stdio.h>
#include <string.h>

#define SIZE 32

char grid[SIZE * SIZE];
char next_grid[SIZE * SIZE];

int wrap(int n) { return (n + SIZE) % SIZE; }

int g(int x, int y) { return grid[y * SIZE + x]; }
void gs(int x, int y, int v) { grid[y * SIZE + x] = v; }
void ns(int x, int y, int v) { next_grid[y * SIZE + x] = v; }

int count_neighbors(int x, int y) {
    int count = 0;
    for (int dy = -1; dy <= 1; dy++) {
        for (int dx = -1; dx <= 1; dx++) {
            if (dx == 0 && dy == 0) continue;
            count += g(wrap(x + dx), wrap(y + dy));
        }
    }
    return count;
}

void step_cell(int x, int y) {
    int n = count_neighbors(x, y);
    int alive = g(x, y);
    int next;
    if (alive) {
        next = (n == 2 || n == 3) ? 1 : 0;
    } else {
        next = (n == 3) ? 1 : 0;
    }
    ns(x, y, next);
}

void step(void) {
    for (int y = 0; y < SIZE; y++) {
        for (int x = 0; x < SIZE; x++) {
            step_cell(x, y);
        }
    }
    memcpy(grid, next_grid, SIZE * SIZE);
}

int count_alive(void) {
    int count = 0;
    for (int i = 0; i < SIZE * SIZE; i++) {
        count += grid[i];
    }
    return count;
}

void init_glider(void) {
    memset(grid, 0, SIZE * SIZE);
    gs(2, 1, 1); gs(3, 2, 1); gs(1, 3, 1); gs(2, 3, 1); gs(3, 3, 1);
}

int main(void) {
    init_glider();
    for (int i = 0; i < 1000; i++) step();
    printf("%d\n", count_alive());
    return 0;
}

File: compiler/bench/life.fs
\ expected: 5
\ Game of Life - run simulation steps

32 constant SIZE
create grid SIZE SIZE * allot
create next-grid SIZE SIZE * allot

: idx ( x y -- addr ) SIZE * + ;
: g@ ( x y -- val ) idx grid + c@ ;
: g! ( val x y -- ) idx grid + c! ;
: n@ ( x y -- val ) idx next-grid + c@ ;
: n! ( val x y -- ) idx next-grid + c! ;

: wrap ( n -- n' ) SIZE + SIZE mod ;

: count-neighbors ( x y -- count )
  0
  3 0 do 3 0 do
    i 1 = j 1 = and if else
      4 pick i + 1- wrap
      4 pick j + 1- wrap
      g@ +
    then
  loop loop
  nip nip ;

: step-cell ( x y -- )
  2dup count-neighbors   \ x y neighbors
  2 pick 2 pick g@       \ x y neighbors alive?
  if
    dup 2 < if drop 0 else 3 > if 0 else 1 then then
  else
    3 = if 1 else 0 then
  then
  rot rot n! ;

: step ( -- )
  SIZE 0 do SIZE 0 do
    j i step-cell
  loop loop
  SIZE SIZE * 0 do
    next-grid i + c@ grid i + c!
  loop ;

: count-alive ( -- n )
  0 SIZE SIZE * 0 do grid i + c@ + loop ;

: init-glider ( -- )
  SIZE SIZE * 0 do 0 grid i + c! loop
  1 1 2 g!  1 2 3 g!  1 3 1 g!  1 3 2 g!  1 3 3 g! ;

: main
  init-glider
  1000 0 do step loop
  count-alive . cr ;
